### PR TITLE
28 create photo app

### PR DESCRIPTION
**Pull request by oshtohryn**
## TRELLO

* [Main Trello ticket](https://trello.com/c/bC4g4S1z/28-create-photo-app)

## Code reviewers

- [ ] @Ashiato
- [ ] @black-art
- [x] @hrytskivr
- [x] @KarnaushenkoVO
- [ ] @PanLeon
- [ ] @spycheast
- [x] @yuriiZherebukh

### Second Level Review

- [ ] @byshovets
- [ ] @lhalam

## Summary of change

Added methods for trip and checkpoint selects
Added time when updated object

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=90%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

File: myTrip/photo/models.py [hunk 1]
```diff
@@ -42,27 +50,75 @@ def get_by_trip_id(trip_id):
...
 
-    def update(self, description=None):
-        """Updating photo model."""
-        if description:
+    def update(self, user_id, description=None):
+        """Updating photo description."""
+        if description and user_id:
             self.description = description
+            self.updated_at = datetime.now()
+            print(datetime.now())
```
Comment (oshtohryn) [inline myTrip/photo/models.py]: Remove print()

File: myTrip/photo/models.py [hunk 2]
```diff
@@ -1,25 +1,33 @@
...
 class Photo(models.Model):
     """
     Photo
     :argument id: int - auto generated primary key
     :argument src: url - photo source link
-    :argument user_id: int - ToDo foreign key to User model id
-    :argument trip_id: int - Todo foreign key to Trip model id
-    :argument checkpoint_id: int - ToDo foreign to Checkpoint model id
-    :argument description: str - description to photo.
+    :argument user: int - foreign key to User model id
```
Comment (lhalam) [inline myTrip/photo/models.py]: int or User?

File: myTrip/photo/models.py [hunk 3]
```diff
@@ -42,27 +50,57 @@ def get_by_trip_id(trip_id):
...
             trip_id (int): trip id.
         Returns:
-            QuerySet<Photo>: QuerySet of Photo.
+            QuerySet<Photos>: QuerySet of Photos.
         """
         photos = Photo.objects.filter(trip_id=trip_id)
         return photos
 
     @staticmethod
+    def get_by_photo_and_user_id(photo_id, user_id):
```
Comment (lhalam) [inline myTrip/photo/models.py]: get_by_photo_and_user_id and get_by_trip_id_and_checkpoint_id merge to one method

File: myTrip/photo/models.py [hunk 4]
```diff
@@ -42,27 +50,57 @@ def get_by_trip_id(trip_id):
...
+            checkpoint_id (int): checkpoint id.
+        Returns:
+            QuerySet<Photos>: QuerySet of Photos.
+        """
+        photos = Photo.objects.filter(trip_id=trip_id, checkpoint_id=checkpoint_id)
+        return photos
+
+
+    @staticmethod
     def create(src, user_id, trip_id=None, checkpoint_id=None, description=None):
```
Comment (lhalam) [inline myTrip/photo/models.py]: trip_id=None, checkpoint_id=None =>trip=None, checkpoint=None

File: myTrip/photo/models.py [hunk 5]
```diff
@@ -42,27 +50,57 @@ def get_by_trip_id(trip_id):
...
         """ Creating photo model, and returns created object"""
         photo = Photo()
         photo.src = src
-        photo.user_id = user_id
-        photo.trip_id = trip_id
-        photo.checkpoint_id = checkpoint_id
+        if trip_id:
+            photo.trip_id = Trip.get_by_id(trip_id).id
```
Comment (lhalam) [inline myTrip/photo/models.py]: deplete get methods

File: myTrip/photo/models.py [hunk 6]
```diff
@@ -42,27 +50,57 @@ def get_by_trip_id(trip_id):
...
+            photo.checkpoint_id = Checkpoint.get_by_id(checkpoint_id).id
+        photo.user = CustomUser.get_by_id(user_id)
         photo.description = description
         photo.save()
         return photo
 
-    def update(self, description=None):
-        """Updating photo model."""
-        if description:
+    def update(self, user_id, description=None):
```
Comment (lhalam) [inline myTrip/photo/models.py]: delete user_id

**Review by lhalam — COMMENTED** [review]
ToDo

File: myTrip/photo/models.py [hunk 7]
```diff
@@ -1,25 +1,33 @@
 """Module contain photo model class and methods."""
 
-from django.core.exceptions import ObjectDoesNotExist
+from datetime import datetime
+
 from django.db import models
+from checkpoint.models import Checkpoint
```
Comment (hrytskivr) [inline myTrip/photo/models.py]: separate different import source with 1 space

File: myTrip/photo/views.py [hunk 8]
```diff
@@ -1,40 +1,50 @@
...
-        photo = Photo.get_by_id(photo_id)
-        if not photo:
-            return HttpResponse(status=404)
-        photo = photo.to_dict()
-        return JsonResponse(photo, status=200, safe=False)
+    def get(self, request, trip_id=None, checkpoint_id=None):
+        """GET request handler. Can return photo/photos ordered by given args."""
+        if trip_id and checkpoint_id:
+            photos = Photo.get_by_trip_id_and_checkpoint_id(trip_id, checkpoint_id)
+        photos = Photo.get_by_trip_id(trip_id)
```
Comment (hrytskivr) [inline myTrip/photo/views.py]: what if we have no photos given???
missing return in 1st if case + consider adding another if statement for better readability:
```
if trip_id and checkpoint_id:
    photos = Photo.get_by_trip_id_and_checkpoint_id(trip_id, checkpoint_id)
    data = [photo.to_dict() for photo in photos]
    return JsonResponse(data, status=200, safe=False)
if trip_id:
    photos = Photo.get_by_trip_id(trip_id)
    data = [photo.to_dict() for photo in photos]
    return JsonResponse(data, status=200, safe=False)
```

File: myTrip/photo/views.py [hunk 9]
```diff
@@ -1,40 +1,50 @@
...
+    def post(self, request, trip_id=None, checkpoint_id=None):
+        """POST request hangler.Creating a new photo object and return status 201("created")"""
         post_data = json.loads(request.body.decode('utf-8'))
-        photo = Photo.create(**post_data).to_dict()
-        return JsonResponse(photo, status=201)
+        if trip_id and checkpoint_id:
+            photo = Photo.create(trip_id=trip_id, checkpoint_id=checkpoint_id, **post_data)
+            data = photo.to_dict()
+            return JsonResponse(data, status=201)
+        photo = Photo.create(trip_id=trip_id, **post_data)
```
Comment (hrytskivr) [inline myTrip/photo/views.py]: another if

**Review by hrytskivr — COMMENTED** [review]
Read comments :)

File: myTrip/photo/views.py [hunk 10]
```diff
@@ -1,38 +1,56 @@
...
 import json
+
 from django.views.generic.base import View
 from django.http import JsonResponse, HttpResponse
+
+from checkpoint.models import Checkpoint
+from registration.models import CustomUser
+from trip.models import Trip
 from .models import Photo
+from datetime import datetime
```
Comment (KarnaushenkoVO) [inline myTrip/photo/views.py]: you should add it above, where you import modules from standart django\python packages

File: myTrip/photo/views.py [hunk 11]
```diff
@@ -1,38 +1,56 @@
...
+                return HttpResponse(status=404)
+            data = [photo.to_dict() for photo in photos]
+            return JsonResponse(data, status=200, safe=False)
         photo = Photo.get_by_id(photo_id)
         if not photo:
             return HttpResponse(status=404)
-        photo = photo.to_dict()
-        return JsonResponse(photo, status=200, safe=False)
+        data = photo.to_dict()
+        return JsonResponse(data, status=200, safe=False)
```
Comment (KarnaushenkoVO) [inline myTrip/photo/views.py]: safe=False is not necessary here

File: myTrip/photo/models.py [hunk 12]
```diff
@@ -28,41 +38,45 @@ def get_by_id(photo_id):
...
         photo.save()
         return photo
 
-    def update(self, description=None):
-        """Updating photo model."""
+    def update(self, description):
+        """Updating photo description."""
         if description:
             self.description = description
+            self.updated_at = datetime.now()
```
Comment (lhalam) [inline myTrip/photo/models.py]: No need to update this field!
`updated_at = models.DateTimeField(auto_now=True, null=True, editable=True)`
https://docs.djangoproject.com/en/1.11/ref/models/fields/#django.db.models.DateField.auto_now

File: myTrip/photo/models.py [hunk 13]
```diff
@@ -1,25 +1,35 @@
...
 
     src = models.URLField(max_length=200)
-    user_id = models.IntegerField()
-    trip_id = models.IntegerField(null=True)
-    checkpoint_id = models.IntegerField(null=True)
+    user = models.ForeignKey(CustomUser, null=True)
+    trip = models.ForeignKey(Trip, null=True)
+    checkpoint = models.ForeignKey(Checkpoint, null=True)
     description = models.TextField(null=True)
+    created_at = models.DateTimeField(auto_now_add=True, null=True)
```
Comment (lhalam) [inline myTrip/photo/models.py]: This field must be required!

File: myTrip/photo/models.py [hunk 14]
```diff
@@ -1,25 +1,35 @@
...
+    updated_at = models.DateTimeField(auto_now=True, null=True, editable=True)
```
Comment (lhalam) [inline myTrip/photo/models.py]: This field must be required!

File: myTrip/photo/models.py [hunk 15]
```diff
@@ -1,25 +1,35 @@
...
-    :argument description: str - description to photo.
+    :argument user: - foreign key to User model
+    :argument trip: - foreign key to Trip model
+    :argument checkpoint: - foreign to Checkpoint model
+    :argument description: str - description to photo
+    :argument created_at: date - time when created
+    :argument updated_at: date - time when updated.
     """
 
     src = models.URLField(max_length=200)
```
Comment (lhalam) [inline myTrip/photo/models.py]: Max_length is used by default of 200, should be described if the value is different from the default.
Https://docs.djangoproject.com/en/1.11/ref/models/fields/#urlfield

File: myTrip/photo/models.py [hunk 16]
```diff
@@ -28,41 +38,45 @@ def get_by_id(photo_id):
...
-            trip_id (int): trip id.
+            trip_id (int): trip id
+            checkpoint_id (int): checkpoint id.
         Returns:
-            QuerySet<Photo>: QuerySet of Photo.
+            QuerySet<Photos>: QuerySet of Photos.
         """
-        photos = Photo.objects.filter(trip_id=trip_id)
-        return photos
+        try:
```
Comment (lhalam) [inline myTrip/photo/models.py]: You do not need to use `try` because the filter does not raise an error

File: myTrip/photo/models.py [hunk 17]
```diff
@@ -72,17 +86,21 @@ def to_dict(self):
...
-            "user_id": self.user_id,
-            "trip_id": self.trip_id,
-            "checkpoint_id": self.checkpoint_id,
-            "description": self.description
+            "user": self.user.id,
+            "trip_id": self.trip.id if self.trip else None,
+            "checkpoint_id": self.checkpoint.id if self.checkpoint else None,
+            "description": self.description,
+            "created": self.created_at,
+            "last updated": self.updated_at
```
Comment (lhalam) [inline myTrip/photo/models.py]: It's a bad practice to use a space in the names of the keys name.
For fields created_at and updated_at use keys with the same names

File: myTrip/photo/views.py [hunk 18]
```diff
@@ -1,38 +1,56 @@
...
 class PhotoView(View):
     """Class that handle HTTP requests."""
 
-    def get(self, request, photo_id):
-        """GET request handler.Get photo by photo_id"""
+    def get(self, request, trip_id=None, checkpoint_id=None, photo_id=None):
+        """GET request handler. Can return photo/photos ordered by given args."""
+        if not photo_id:
+            photos = Photo.filter(trip_id, checkpoint_id)
+            if not photos:
```
Comment (lhalam) [inline myTrip/photo/views.py:22]: Is this really a 404 error?

File: myTrip/photo/views.py [hunk 19]
```diff
@@ -1,38 +1,56 @@
...
+        return JsonResponse(data, status=200)
 
-    def post(self, request):
-        """POST request hangler.Creating a new photo object"""
+    def post(self, request, trip_id=None, checkpoint_id=None):
+        """POST request hangler.Creating a new photo object and return status 201("created")"""
         post_data = json.loads(request.body.decode('utf-8'))
-        photo = Photo.create(**post_data).to_dict()
-        return JsonResponse(photo, status=201)
+        user = CustomUser.get_by_id(post_data["user"])
```
Comment (lhalam) [inline myTrip/photo/views.py]: After the implementation of AUTH, you will use request.user.id, you do not need to transfer in request.body

File: myTrip/photo/views.py [hunk 20]
```diff
@@ -1,38 +1,56 @@
...
+        trip = Trip.get_by_id(trip_id)
+        checkpoint = Checkpoint.get_by_id(checkpoint_id)
+        photo = Photo.create(trip=trip, checkpoint=checkpoint,
+                             user=user, src=post_data["src"], description=post_data["description"])
+        data = photo.to_dict()
+        return JsonResponse(data, status=201)
 
-    def put(self, request, photo_id):  # pylint: disable=no-self-use
+    def put(self, request, trip_id=None, checkpoint_id=None, photo_id=None):  # pylint: disable=unused-argument,no-self-use
```
Comment (lhalam) [inline myTrip/photo/views.py]: The PUT method should only have a photo_id

File: myTrip/photo/views.py [hunk 21]
```diff
@@ -1,38 +1,56 @@
...
         if not photo:
             return HttpResponse(status=404)
-        update_data = json.loads(request.body.decode('utf-8'))
         photo.update(**update_data)
-        return HttpResponse(status=200)
+        data = photo.to_dict()
+        return JsonResponse(data, status=200)
 
-    def delete(self, request, photo_id):  # pylint: disable=unused-argument,no-self-use
+    def delete(self, request, trip_id=None, checkpoint_id=None, photo_id=None):  # pylint: disable=unused-argument,no-self-use
```
Comment (lhalam) [inline myTrip/photo/views.py]: The delete method should only have a photo_id

**Review by lhalam — COMMENTED** [review]
ToDo

**Review by hrytskivr — APPROVED** [review]
Everything looks fine, Merging!

**Pull request merged**